sw/device/lib/crypto/aes_gcm/aes_gcm.c

```diff
@@ -354,3 +374,106 @@ aes_error_t aes_gcm_ghash(const aes_block_t *hash_subkey,
+ */
+static aes_error_t aes_single_block(const aes_params_t params,
+                                    const aes_block_t *input,
+                                    aes_block_t *output) {
+  aes_error_t err = aes_begin(params);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_update(output, input);
+  if (err != kAesOk) {
+    return err;
+  }
+  err = aes_end();
+  if (err != kAesOk) {
+    return err;
+  }
+  return kAesOk;
+}
+
+/**
+ * Implements the GCTR function as specified in SP800-38D, section 6.5.
+ *
+ * The block cipher is fixed to AES. Note that the GCTR function is a modified
+ * version of the AES-CTR mode of encryption.
+ *
+ * Input must be less than 2^32 blocks long; that is, `len` < 2^36, since each
+ * block is 16 bytes.
+ *
+ * @param key_len length of the AES key
+ * @param key_shares key, expressed in two shares
+ * @param icb Initial counter block, 128 bits
+ * @param len Number of bytes for input and output
+ * @param input Pointer to input buffer (may be NULL if `len` is 0)
+ * @param output Pointer to output buffer (same size as input, may be the same
+ * buffer)
+ */
+aes_error_t aes_gcm_gctr(const aes_key_len_t key_len, uint32_t *key_shares[2],
+                         const aes_block_t *icb, const size_t len,
+                         const uint8_t *input, uint8_t *output) {
+  // If the input is empty, the output must be as well. Since the output length
+  // is 0, simply return.
+  if (len == 0) {
+    return kAesOk;
+  }
+
+  // NULL pointers are not allowed for nonzero input length.
+  if (input == NULL || output == NULL) {
+    return kAesInternalError;
+  }
+
+  // Prepare AES parameters for AES-CTR encryption. Encrypting a single block B
+  // with these parameters will return AES_K(IV) ^ B, where AES_K is the core
+  // AES block cipher encryption operation using key K.
+  aes_params_t aes_ctr_params = {
+      .encrypt = true,
+      .mode = kAesCipherModeCtr,
+      .key_len = key_len,
+      .key = {key_shares[0], key_shares[1]},
+      .iv = {0},
+  };
+  memcpy(aes_ctr_params.iv, icb->data, kAesBlockNumBytes);
+
+  // Calculate the number of blocks needed to represent the input. Since we
+  // checked for empty input above, nblocks must be at least 1.
+  size_t nblocks = get_nblocks(len);
+  for (size_t i = 0; i < nblocks; ++i) {
+    // Retrieve the next block of input. All blocks are full-size except for
+    // the last block, which may be partial. If the block is partial, the input
+    // data will be padded with zeroes.
+    aes_block_t block_in = {.data = {0}};
+    size_t nbytes = kAesBlockNumBytes;
+    if ((i == nblocks - 1) && (get_block_offset(len) != 0)) {
+      // Last block is partial; copy over only the bytes that exist.
+      nbytes = get_block_offset(len);
+      memcpy(block_in.data, &input[i * kAesBlockNumBytes], nbytes);
+    } else {
+      // This block is a full block.
+      memcpy(block_in.data, &input[i * kAesBlockNumBytes], kAesBlockNumBytes);
+    }
```

Comment:
I'm not sure if `aes_block_t block_in = {.data = {0}};` is optimized by the compiler but if not, do we want to memset each block to 0? You may want to consider something like:
```
if (i == (nblocks - 1)) {
    size_t last_block_num_bytes = get_last_block_num_bytes(len);
    memcpy(block_in.data, &input[i * kAesBlockNumBytes], last_block_num_bytes);
    if (last_block_num_bytes < kAesBlockNumBytes) {
        memset(block_in.data + last_block_num_bytes, 0, kAesBlockNumBytes - last_block_num_bytes);
    }
}
```

Reply:
Good idea, thanks! I'll implement this and the name change in your other comment in the next PR.

Reply:
SG, thanks a lot for this PR and apologies for the late review.